iohk-monitoring/src/Cardano/BM/Backend/Switchboard.lhs

```diff
@@ -226,29 +244,55 @@ instance (FromJSON a, ToJSON a) => IsBackend Switchboard a where
         -- raises an exception, that exception will be re-thrown in the current
         -- thread, wrapped in ExceptionInLinkedThread.
         Async.linkOnly (not . isBlockedIndefinitelyOnSTM) dispatcher
-        putMVar sbref $ SwitchboardInternal {
-                            sbQueue = q,
-                            sbDispatch = dispatcher,
-                            sbLogBuffer = logbuf,
-                            sbLogBE = katipBE,
-                            sbBackends = bs }
+
+        -- Modify the internal state of the switchboard, the switchboard
+        -- is now running.
+        putMVar sbref $ SwitchboardInternal 
+            { sbQueue = q
+            , sbDispatch = dispatcher
+            , sbLogBuffer = logbuf
+            , sbLogBE = katipBE
+            , sbBackends = bs
+            , sbRunning = SwitchboardRunning
+            }
 
         return sb
 
     unrealize switchboard = do
-        let clearMVar :: MVar some -> IO ()
+        -- How is this "clearing" anything?
+        let clearMVar :: MVar (SwitchboardInternal a) -> IO ()
             clearMVar = void . tryTakeMVar
 
+        -- This is kind of weird. This function is generally used for
+        -- the modification of the internal state, not breaking the encapsulation.
         (dispatcher, queue) <- withMVar (getSB switchboard) (\sb -> return (sbDispatch sb, sbQueue sb))
```

Comment:
right. this should better get a lock on the MVar, send the kill pill and then clean up. or, was the idea to give all the backends a last chance to use the switchboard?

Reply:
Exactly, this should be done _inside_ the `withMVar`.
I don't see how this gives all the backends any chance in either case.